### ISSUE TITLE
Upgrading from V1 > V2 drops + creates new API

### ISSUE THREAD

**Issue by manwaring** [sid88in/serverless-appsync-plugin#617]
First of all thank you so much for creating + maintaining this plugin, it's been so helpful!

I went through the steps to upgrade from V1 to V2 and was able to deploy successfully, however during the deployment the prior Appsync API is destroyed and a new one is created.

I found this issue about [v2.3.0 dropping + recreating the API](https://github.com/sid88in/serverless-appsync-plugin/issues/607) but I don't think that's related to what I'm seeing, as I'm not making any changes to my schema or configurations other than those [required for going from V1 to V2](https://github.com/sid88in/serverless-appsync-plugin/blob/master/doc/upgrading-from-v1.md).

Is this a known issue, and/or did I miss a step in the upgrade guide?

**Comment by bboure** [comment]
Hi,

In order to understand what is going on we'd need to check a few things

1. Would you be able to share your v1 and v2 serverless.yml (the relevant parts. i.e. AppSync)
2. Check the generated CloudFormation template from v1 and v2 and look for differences (e.g. different logical id).

**Comment by manwaring** [comment]
It looks like the Logical ID does in fact change from v1 to v2

v1 Logical ID for GraphQLApi
```JSON
  "AppsynctestGraphQlApi": {
    "Type": "AWS::AppSync::GraphQLApi",
    "Properties": {
      "Name": "appsync-test",
      "AuthenticationType": "AMAZON_COGNITO_USER_POOLS",
      "AdditionalAuthenticationProviders": [
        {
          "AuthenticationType": "AWS_IAM"
        }
      ],
      "UserPoolConfig": {
        "AwsRegion": "us-east-1",
        "UserPoolId": {
          "Ref": "UserPool"
        },
        "DefaultAction": "ALLOW"
      },
      "LogConfig": {
        "CloudWatchLogsRoleArn": {
          "Fn::GetAtt": [
            "AppsynctestGraphQlApiCloudWatchLogsRole",
            "Arn"
          ]
        },
        "FieldLogLevel": "ERROR",
        "ExcludeVerboseContent": false
      },
      "XrayEnabled": false
    }
  }
```

v2 Logical ID for GraphQLApi
```JSON
  "GraphQlApi": {
    "Type": "AWS::AppSync::GraphQLApi",
    "Properties": {
      "Name": "appsync-test",
      "XrayEnabled": false,
      "AuthenticationType": "AMAZON_COGNITO_USER_POOLS",
      "UserPoolConfig": {
        "AwsRegion": "us-east-2",
        "UserPoolId": {
          "Ref": "UserPool"
        },
        "DefaultAction": "ALLOW"
      },
      "AdditionalAuthenticationProviders": [
        {
          "AuthenticationType": "AWS_IAM"
        }
      ],
      "LogConfig": {
        "CloudWatchLogsRoleArn": {
          "Fn::GetAtt": [
            "GraphQlApiLogGroupRole",
            "Arn"
          ]
        },
        "FieldLogLevel": "ERROR",
        "ExcludeVerboseContent": false
      }
    }
  }
```

**Comment by bboure** [comment]
It looks like you were using the multiple APIs feature? (maybe without knowing it). 

in v1, was your `custom.appSync` config passed as an array?

**Comment by manwaring** [comment]
yeah it was, not intentionally (I didn't have multiple) just because that's the way I must have set it up initially (and in this example app to reproduce)

**Comment by bboure** [comment]
OK. So that's the issue. 

Unfortunately, there is no easy way out of this as of today. 

Are you using a custom domain by any chance? 
That would be the simplest solution: deploy a new API and point your custom domain to it, then remove the old one. 

Another workaround would be to add an escape hatch by allowing a custom LogicalId for the main API resource, and maybe other things like API keys, etc.  But this is trickier

**Comment by manwaring** [comment]
shoot, I wish I could go back in time a year ago and configure it differently 😅

I'd like to avoid using custom domains as i've had config issues with it in the past that have been difficult to resolve (so prefer to use generated domain if at all possible).

Would you be open to a PR that exposes an optional parameter for naming the appsync logical id?

**Comment by manwaring** [comment]
This is a simple implementation with updated tests in case seeing it helps

https://github.com/sid88in/serverless-appsync-plugin/pull/618

**Comment by terrancej** [comment]
Is anyone looking at this, im also in the same boat

**Comment by bboure** [comment]
Sorry, I'll start looking at this starting next week.
Thank you for your patience

**Comment by eddiesmithjr** [comment]
I will close my issue as this is the root of the same issue I'm having [here](https://github.com/sid88in/serverless-appsync-plugin/issues/620)

**Comment by sid88in** [comment]
This is a known consequence of the v1→v2 change: the underlying CloudFormation logical resources differ, so the API is replaced rather than updated in place. There isn't an in-place upgrade path. Documenting as a known limitation and closing — see #594 for the related "adopt an existing API by id" work.
